### DISCUSSION TITLE
Integration of interactive measure tool after knickpoint extraction

### DISCUSSION THREAD

**Discussion by jimpaulobautista** [wschwanghart/topotoolbox#36] [Ideas]
I am very impressed with the capability of TopoToolbox in automating every process in hydrological analysis. Just an idea, after extracting the knickpoints and plotting them on the maps, can the GUI of the measure tool be activated to derive slope measurements between the knickpoints (represented as circles)? I was able to load it by adjusting the code; however, it freezes after zooming in to the DEM. 
![image](https://github.com/wschwanghart/topotoolbox/assets/108282093/e72d6282-2a84-4fe9-bf9c-4a3fb85d7cbd)

**Comment by wschwanghart** [comment]
Hi, good point. Actually, the measure-app was not foreseen to do this, and perhaps, it might be better to accomplish measuring between knickpoints using the tools available in the class PPS.

For example, you might have a STREAMobj S and a structure array kp calculated with knickpointfinder, then you can do the following:
```
P = PPS(S,'pp',kp.IXgrid,'z',DEM);
d = pointdistances(P);
```
This will give you the distances between all knickpoints along the the fluvial network. ```pointdistances``` also has some options such as 'val', which lets you for example supply your own distance values (e.g. chi obtained with chitransform).

Let me know if this helps.
Cheers, Wolfgang

**Reply by jimpaulobautista** [reply]
This actually works, thanks Wolfgang! I think the remaining problem here is how to compute for the elevation differences between knickpoints then I can easily work on the slope calculation from these values.

Actually I found a way to execute this on Matlab but it is already beyond the environment of TopoToolbox. I used Mapping Toolbox to modify the attribute table of the knickpoints shapefile I exported using PPS. I followed these steps:

- convert the knickpoint shapefile into a modifiable table using `struct2table `
- since the shapewrite function stores data such as upstream distance (distup) and elevation (elev), I can exploit these data to compute for the slope between knickpoints
- the catch is, the rows of the table are randomly sorted (probably according to increasing/decreasing values of dz) so I sorted the rows according to increasing distup values
- I computed for the difference between two consecutive cells in the distup column to compute for the change in distance
- Similarly, I also created a command to compute for the change in elevation between two consecutive cells in the elev column
- I divided the values I got from these operations to get the slope values between knickpoints
- convert the table back to a shapefile then export the shapefile

This solution is quite artificial in terms of automating the process, but since the functions of TopoToolbox are still limited, I found this way to go. I also attempted to modify the established commands in TopoToolbox but all of them did not work out. Still, I am hopeful that this can be integrated in the functions of TopoToolbox in the future, because the slope differences between knickpoints can give additional insights in interpreting the tectonics and geomorphology of the area.

That's all, thanks.

**Reply by wschwanghart** [reply]
Hi, there are certainly multiple ways to do this, and I am glad you have a procedure that works, although the random sorting seems strange to me. Alternatively, you could try to derive the directed adjacency matrix of the points and work with this one. There's no function to calculate the adjacency matrix, but you can get it from the output of `pointdistances`. Here's how:
``` 
DEM = GRIDobj('srtm_bigtujunga30m_utm11.tif');
FD = FLOWobj(DEM);
S = STREAMobj(FD,'minarea',1000);
S = klargestconncomps(S,1);
 
[~,kp] = knickpointfinder(S,DEM,'tol',20,'plot',false,...
    'split',false,'verbose',false);

P = PPS(S,"PP",kp.IXgrid,'z',DEM);
D = pointdistances(P,'type','digraph','output','matrix',...
    'direction','down');

% Set distances on the main diagonal to inf
D(speye(size(D))>0) = inf;

% Find locations of minimum distances
m = min(D,[],2);
A = D == m & ~isinf(D);

% A is now a directed adjacency matrix of the nearest downstream points
[ix,ixc] = find(A);
% Distance of points
dp = getmarks(P,P.S.distance);
% Elevation of points
zp = getmarks(P,DEM);
% Interpoint distances and interpoint elevation differences
dpd = inf(size(dp));
zpd = inf(size(dp));
% The distance of each point to its downstream neighbor
dpd(ix) = dp(ix)-dp(ixc); 
% The elevation difference of each point to its downstream neighbor
zpd(ix) = zp(ix)-zp(ixc);
% In dpd and zpd, values will be inf for points that do not have any
% downstream neighbor

% Check visually, whether you connected the right points.
plot(P)
hold on
gplot(A,[kp.x kp.y])

``` 
So, hope this works.

**Reply by jimpaulobautista** [reply]
Hi Wolfgang.

I've been away for a while so apologies for the delayed feedback.

I tried the workflow you have suggested and amazingly it works. I just added a separate command for the computation of the slope and exported the values together with other parameters using the shapewrite function of PPS. 
![image](https://github.com/wschwanghart/topotoolbox/assets/108282093/a597e33d-90fd-4a55-b301-29b1fbae2f05)

I also compared the slope values I derived using the workflow I discovered (left)  vs. the one you recommended (right). Both of these methods arrive at the same slope values. 

![image](https://github.com/wschwanghart/topotoolbox/assets/108282093/b2193998-a4e5-4ab8-8355-fc72cd40a541)
![image](https://github.com/wschwanghart/topotoolbox/assets/108282093/63c83c45-e8ac-433d-86e5-c10b9f8212ab)

I am happy that I finally got the solution I needed for this particular parameter. Thanks Wolfgang for your patience and guidance in maximizing the applications of TopoToolbox. 

Best, 

Jim

**Reply by wschwanghart** [reply]
Hi Jim,
thanks for your feedback! I very glad that the approach worked out. 
Best, Wolfgang
